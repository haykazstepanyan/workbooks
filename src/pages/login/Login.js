import { useState, useEffect, useCallback } from "react";
import styles from "./login.module.scss";
import { useDispatch, useSelector } from "react-redux";
import { userSignIn } from "../../redux/auth/actionCreators";
import { useHistory } from "react-router-dom";
import { selectUserData, selectAuthError } from "../../redux/auth/selectors";
import { Input, Button } from "../../components/inputs";

function Login() {
	const [email, setEmail] = useState("");
	const [password, setPassword] = useState("");
	const [errors, setErrors] = useState({
		email: "",
		password: "",
		others: "",
	});

	const dispatch = useDispatch();
	const history = useHistory();
	const userData = useSelector(selectUserData);
	const authError = useSelector(selectAuthError);

	useEffect(() => {
		if (authError) {
			const { code, message } = authError;
			if (code.includes("email")) {
				setErrors({
					email: message,
					password: "",
					others: "",
				});
			} else if (code.includes("password")) {
				setErrors({
					email: "",
					password: message,
					others: "",
				});
			} else {
				setErrors({
					email: "",
					password: "",
					others: message,
				});
			}
		}
	}, [authError]);

	const handleEmail = useCallback((e) => {
		setEmail(e.target.value);
	}, []);

	const handlePassword = useCallback((e) => {
		setPassword(e.target.value);
	}, []);

	function handleSubmit(e) {
		e.preventDefault();
		dispatch(userSignIn({ email, password }));
	}

	if (userData?.id) {
		history.push("/");
		return null;
	} else {
		return (
			<div className={styles.login}>
				<form
					onSubmit={handleSubmit}
					noValidate
					autoComplete="off"
					className={styles.login__form}
				>
					<Input
						type="email"
						label="Email"
						id="email"
						value={email}
						handleChange={handleEmail}
						hasError={!!errors.email}
						errorMessage={errors.email}
					/>
					<Input
						type="password"
						label="Password"
						id="password"
						value={password}
						handleChange={handlePassword}
						hasError={!!errors.password}
						errorMessage={errors.password}
						autoComplete="on"
					/>
					{errors.others && (
						<div className="errors">
							<p>{errors.others}</p>
						</div>
					)}
					<Button type="submit" disabled={!email || !password} text="Sign in" />
				</form>
			</div>
		);
	}
}

export default Login;
