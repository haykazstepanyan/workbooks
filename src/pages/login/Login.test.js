import { render } from "@testing-library/react";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import Login from "./";

describe("workplace folder components", () => {
	const mockStore = configureStore();
	const store = mockStore({ auth: {} });
	const { getByTestId } = render(
		<Provider store={store}>
			<Login user={{ type: "" }} />
		</Provider>
	);
	const login = getByTestId("login");

	it("should match component snapshot", () => {
		expect(login).toMatchSnapshot();
	});
});
