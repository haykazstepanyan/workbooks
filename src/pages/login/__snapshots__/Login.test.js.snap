// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workplace folder components should match component snapshot 1`] = `
<div
  class="login"
  data-testid="login"
>
  <form
    autocomplete="off"
    class="login__form"
    novalidate=""
  >
    <div
      class="inputs"
      data-testid="wrapper-box"
    >
      <input
        class="inputs__field"
        data-testid="input"
        id="email"
        type="email"
        value=""
      />
      <label
        class="inputs__labels"
        data-testid="label"
        for="email"
      >
        Email
      </label>
    </div>
    <div
      class="inputs"
      data-testid="wrapper-box"
    >
      <input
        autocomplete="on"
        class="inputs__field"
        data-testid="input"
        id="password"
        type="password"
        value=""
      />
      <label
        class="inputs__labels"
        data-testid="label"
        for="password"
      >
        Password
      </label>
    </div>
    
    <div
      class="buttons_centerAlign"
    >
      <button
        class="buttons"
        data-testid="button"
        disabled=""
        type="submit"
      >
        Sign in
      </button>
    </div>
  </form>
</div>
`;
